internal/xds/v3/callbacks_test.go

```diff
@@ -0,0 +1,126 @@
+	envoy_service_discovery_v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	envoy_type_v3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"github.com/sirupsen/logrus"
+	"github.com/sirupsen/logrus/hooks/test"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/genproto/googleapis/rpc/code"
+	"google.golang.org/genproto/googleapis/rpc/status"
+)
+
+func TestLogDiscoveryRequestDetails(t *testing.T) {
+	log, logHook := test.NewNullLogger()
+	tests := map[string]struct {
+		discoveryReq    *envoy_service_discovery_v3.DiscoveryRequest
+		expectedLogMsg  string
+		expectedLogData logrus.Fields
+	}{
+		"request with node info": {
+			discoveryReq: &envoy_service_discovery_v3.DiscoveryRequest{
+				VersionInfo:   "req-version",
+				ResponseNonce: "resp-nonce",
+				ResourceNames: []string{"some", "resources"},
+				TypeUrl:       "some-type-url",
+				Node: &envoy_config_core_v3.Node{
+					Id: "node-id",
+					UserAgentVersionType: &envoy_config_core_v3.Node_UserAgentBuildVersion{
+						UserAgentBuildVersion: &envoy_config_core_v3.BuildVersion{
+							Version: &envoy_type_v3.SemanticVersion{
+								MajorNumber: 9,
+								MinorNumber: 8,
+								Patch:       7,
+							},
+						},
+					},
+				},
+			},
+			expectedLogMsg: "handling v3 xDS resource request",
+			expectedLogData: logrus.Fields{
+				"version_info":   "req-version",
+				"response_nonce": "resp-nonce",
+				"resource_names": []string{"some", "resources"},
+				"type_url":       "some-type-url",
+				"node_id":        "node-id",
+				"node_version":   "v9.8.7",
+			},
+		},
+		"request without node info": {
+			discoveryReq: &envoy_service_discovery_v3.DiscoveryRequest{
+				VersionInfo:   "req-version",
+				ResponseNonce: "resp-nonce",
+				ResourceNames: []string{"some", "resources"},
+				TypeUrl:       "some-type-url",
+			},
+			expectedLogMsg: "handling v3 xDS resource request",
+			expectedLogData: logrus.Fields{
+				"version_info":   "req-version",
+				"response_nonce": "resp-nonce",
+				"resource_names": []string{"some", "resources"},
+				"type_url":       "some-type-url",
+			},
+		},
+		"request with error detail": {
+			discoveryReq: &envoy_service_discovery_v3.DiscoveryRequest{
+				VersionInfo:   "req-version",
+				ResponseNonce: "resp-nonce",
+				ErrorDetail: &status.Status{
+					Code:    int32(code.Code_INTERNAL),
+					Message: "error message from request",
+				},
+			},
+			expectedLogMsg: "error message from request",
+			expectedLogData: logrus.Fields{
+				"version_info":   "req-version",
+				"response_nonce": "resp-nonce",
+				"code":           int32(code.Code_INTERNAL),
+			},
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			logDiscoveryRequestDetails(log, tc.discoveryReq)
+			var expectedLogEntry *logrus.Entry
+			for _, le := range logHook.AllEntries() {
+				if le.Message == tc.expectedLogMsg {
+					expectedLogEntry = le
```

Comment:
nit, I think you can break the loop at this point if you've found the expected entry